### ISSUE TITLE
Configurable deploys (rbenv support, etc)

### ISSUE THREAD

**Issue by ryansch** [aws/opsworks-cookbooks#141]
I've hacked in rbenv support in my fork of the cookbooks.  Is it something that I should try to turn into a pull request?  It's working very well for us.  I can support new rubies on the same day of release.

**Comment by githuesch** [comment]
Any reason not to use http://community.opscode.com/cookbooks/rbenv instead?

**Comment by ryansch** [comment]
I'm actually using https://github.com/fnichol/chef-rbenv in my approach.  To get a correctly functioning application with rbenv and bundler, I had to make a number of changes to the way ruby commands are invoked in the opsworks-cookbooks code.  Those changes would be the focus of a PR.

**Comment by githuesch** [comment]
So have you forked Fletcher's cookbook, or are you wrapping it?

**Comment by ryansch** [comment]
I had forked it only to apply in progress PRs which I see have now been merged.

Let's take this up a notch.  Here's the compare view: https://github.com/ace-cookbooks/opsworks-cookbooks/compare/aws:release-chef-11.10...configure-deploy

**Comment by ryansch** [comment]
My stack json includes:

``` json
{
  "opsworks": {
    "rails": {
      "bundler_binary": "/usr/local/rbenv/shims/bundle",
      "environment": {
        "RBENV_ROOT": "/usr/local/rbenv",
        "PATH": "/usr/local/rbenv/shims:/usr/local/rbenv/bin:/bin:/usr/bin"
      }
    }
  }
}
```

**Comment by ryansch** [comment]
I have another branch where I'm tracking changes to unicorn.  It looks like a related change would have to be picked from there.  See: https://github.com/ace-cookbooks/opsworks-cookbooks/compare/aws:release-chef-11.10...unicorn#diff-25c5d1461af9c0d071313d46b2865b1eR41

**Comment by ryansch** [comment]
I want to gauge interest in these features before I take the time to wrap them up in a PR.  You could summarize the intent behind the changes as "Give me more control over the environment, bundler, and command invocation".  I was then able to use the new control to implement rbenv support with the above stack json and adding some recipes to the run-list.

**Comment by githuesch** [comment]
I see. The reason we're somewhat hesitant with pull requests that aren't just bug fixes is that we actually don't want to provide more and more features that can be brought in otherwise. Especially with the newly added Berkshelf integration, new functionality should be provided through community cookbooks. If you need glue code that adapts community cookbooks to the OpsWorks environment, per-Berkshelf custom cookbooks are the way to go.

However, I like your change. Let me discuss this internally before I get back to you.

**Comment by ryansch** [comment]
Thanks very much @githuesch.  It's a spike in that direction.  Anything that accomplishes the same goal of giving me more control would work as well.

**Comment by ryansch** [comment]
On the subject of using community cookbooks to implement features: It would help greatly if you guys would namespace your cookbooks where they conflict with a community cookbook.

Use case:
I'm currently trying to give our haproxy config an overhaul to use all the new 1.5 stuff (ssl, http-request, stick tables, etc) and the attribute driven stuff in the community cookbook is great.  It's not clear to me how to use it since the name conflicts with your haproxy cookbook which is already in the run-list.

**Comment by githuesch** [comment]
I totally agree on namespacing, that's a known issue.

**Comment by ryansch** [comment]
Thanks for the ack.  :smile:

**Comment by fred** [comment]
@ryansch regarding your comment on haproxy 1.5, is there any news about it?
Haproxy 1.5 would be very welcoming upgrade, it would solve some issues on the rails stack.

With built in SSL support, it means that it will solve the issue with a rails stack where the HTTPS requests dont have x-forwarded-for header. Haproxy 1.4 https requests uses the TCP mode, where there is no way for the rails stack application to get the client IP, it will always be 127.0.0.1

**Comment by awsrequena** [comment]
@fred we now install haproxy per default OS package.

https://launchpad.net/ubuntu/+source/haproxy shows ""haproxy (1.5.3-1~ubuntu14.04.1) trusty-backports".  Maybe if you could add some apt source to your system and install this version.

**Comment by fred** [comment]
@awsrequena I noticed that, and it's great.
still I would have to overwrite the haproxy template to match the new configuration between haproxy 1.4 and 1.5 and to use SSL provided to the application json.

I really would like to fix the issue that on HTTPS traffic the client IP is not forwarded ("x-forwarded-for" header), I have not yet experienced using HAProxy 1.5

I will have a look at it soon on a test stack.
Cheers.

**Comment by awsrequena** [comment]
@ryansch @fred do you guys will be OK with closing this issue?

**Comment by ryansch** [comment]
This one slipped past me.  I'd rather not close this as the original issue was about creating a PR with the code I've written.  I'd like to know if you guys want to see it.

**Comment by ryansch** [comment]
@fred I use yum with a plugin to add s3 repo support but as mentioned earlier it's easier to get 1.5 now that it has gone stable.  As far as the config, it's pretty easy to overwrite it from a custom cookbook.

**Comment by awsrequena** [comment]
@ryansch I like  https://github.com/ace-cookbooks/opsworks-cookbooks/compare/aws:release-chef-11.10...configure-deploy ... but I need to take a deeper look. Please create a PR for this and link it here.  Btw. the " --deployment" is not going to work just like that, as it changes the current behaviour of the bundle execution.  Please also include the extra JSON structure in the PR description.

As for the https://github.com/ace-cookbooks/opsworks-cookbooks/compare/aws:release-chef-11.10...unicorn#diff-25c5d1461af9c0d071313d46b2865b1eR41 ... that is a big one.  This also needs to 
be evaluated and throughly reviewed.  We really must ensure backwards compatibility in current features. Please create a PR for this too.

As for both PR, we need that both are published under the Apache License version 2.0 (http://aws.amazon.com/apache2.0/)
